server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -0,0 +1,148 @@
+        this.fileTransferListener = fileTransferListener;
+        this.exclusionFilter = exclusionFilter;
+    }
+
+    public boolean transferSnapshot(TransferSnapshot transferSnapshot, TranslogTransferListener translogTransferListener)
+        throws IOException {
+        List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+        Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+        try {
+            toUpload.addAll(exclusionFilter.apply(transferSnapshot.getTranslogFileSnapshots()));
+            toUpload.addAll(exclusionFilter.apply(transferSnapshot.getCheckpointFileSnapshots()));
+            final CountDownLatch latch = new CountDownLatch(toUpload.size());
+            LatchedActionListener<TransferFileSnapshot> latchedActionListener = new LatchedActionListener<>(
+                ActionListener.wrap(fileTransferListener::onSuccess, ex -> {
+                    assert ex instanceof FileTransferException;
+                    logger.error(
+                        () -> new ParameterizedMessage(
+                            "Exception during transfer for file {}",
+                            ((FileTransferException) ex).getFileSnapshot().getName()
+                        ),
+                        ex
+                    );
+                    FileTransferException e = (FileTransferException) ex;
+                    fileTransferListener.onFailure(e.getFileSnapshot(), ex);
+                    exceptionList.add(ex);
+                }),
+                latch
+            );
+            toUpload.forEach(
+                fileSnapshot -> transferService.uploadBlobAsync(
+                    fileSnapshot,
+                    remoteBaseTransferPath.add(String.valueOf(fileSnapshot.getPrimaryTerm())),
+                    latchedActionListener
+                )
+            );
+            try {
+                if (latch.await(TRANSFER_TIMEOUT_IN_MILLIS, TimeUnit.MILLISECONDS) == false) {
+                    Exception ex = new TimeoutException("Timed out waiting for transfer of snapshot " + transferSnapshot + " to complete");
+                    exceptionList.forEach(ex::addSuppressed);
+                    throw ex;
+                }
+            } catch (InterruptedException ex) {
+                exceptionList.forEach(ex::addSuppressed);
+                logger.error(() -> new ParameterizedMessage("Time failed for snapshot {}", transferSnapshot), ex);
+                Thread.currentThread().interrupt();
+                throw ex;
+            }
+            if (exceptionList.isEmpty()) {
+                final TransferFileSnapshot transferFileSnapshot = prepareMetadata(transferSnapshot);
+                transferService.uploadBlob(
+                    prepareMetadata(transferSnapshot),
+                    remoteBaseTransferPath.add(String.valueOf(transferFileSnapshot.getPrimaryTerm()))
+                );
```

Comment:
what is the rational for doing sync call here vs doing it async ? In normal cases, the async call will be uploading two files (ckp and tlog), where this is uploading one file (metadata) .

Reply:
We want to upload metadata once we are sure all referenced files are uploaded successfully else we will have to do a rollback if any one of them fails

Reply:
sure, but we are doing this only if `exceptionList.isEmpty()` , which means that they have already been uploaded successfully.

Reply:
Oh since we don't have a non blocking listener support as yet for the overall translog transfer